keps/sig-apps/3541-add-enforcedRollingUpdate-to-statefulset-update-strategies/README.md

```diff
@@ -0,0 +1,923 @@
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+Currently we support two rolling update strategies in statefulSet, one is `OnDelete`, another one is `RollingUpdate`,
+which is the default one. And as designed, when we got a broken state with a statefulSet applying the default `RollingUpdate`
+strategy, we have to repair this manually.
+
+Like we applied a broken statefulSet, and the corresponding pods will stuck in pending with failures. then we apply with
+a fixed yaml file, the stuck pods will not turn to health for security design, see [Forced rollback](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#forced-rollback).
+
+But this is heavily complained by the community, see [issue-67250](https://github.com/kubernetes/kubernetes/issues/67250),
+[issue-60164](https://github.com/kubernetes/kubernetes/issues/60164), [issue-109597](https://github.com/kubernetes/kubernetes/issues/109597) and so on,
+we should provide a escape patch for end-users.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+As described above, pods in a broken state won't turn to health even we apply with a fixed statefulSet yaml file, instead,
+statefulSet controller will wait until we delete the pod manually. I think this was designed because of some security reasons for
+statefulSet generally is used to manage stateful workloads, delete them directly may lead to some unexpected results like
+data loss, this is intolerable.
+
+But not always that case, for example we may deploy a statefulset just for the uniqueness of Pods, automatically delete them will not
+cause any serious problems, so we should treat the rolling-update strategies case by case, and better to hand the choices over to end-users.
+
+From the solutions of community users, we have two choices to avoid this, one is set the `podManagementPolicy` to `Parallel`, see [issue](https://github.com/jenkinsci/helm-charts/issues/688),
+but we will update the pods in parallel. Another one is use a controller and find the broken pods to delete, see [PR](https://github.com/argoproj-labs/argocd-operator/pull/604).
```

Comment:
Pod management policy should update in parallel up to the pod limit.  If you set maxunavailable “1”, that sounds like the same as what this kep proposes?  Or did I miss a key difference in the algorithm you are proposing?

I think I need more detail to understand what you expect to happen - if you could describe a sequence of events that results in a “stuck” scenario (1-N each step) that would be very useful.

Reply:
@smarterclayton  can you look at this issue(https://github.com/kubernetes/kubernetes/issues/67250) which is being referenced in this KEP. The scenario the author is highlighting is when a statefulset goes into a broken state due to a rolling update of a bad image as one example which requires deleting the pod, even after the image is fixed to correct image and was probably done to prevent data loss.  Thoughts ?

Reply:
Correct, we even explicitly explain that case in the [docs](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#forced-rollback), so this was intentional. I'm not saying we shouldn't expand that functionality, but I'd like to hear more during one of sig-apps calls.

Reply:
@soltysh This kind of (not) bugfixing seems strange to me: Just because you [document](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#forced-rollback) an issue (e.g. https://github.com/kubernetes/kubernetes/issues/67250) doesn't make it go away.

The docs don't even mention a workaround to avoid this behaviour. 

"to prevent dataloss" - how do you want to recover the data, if you are not able to replace the pod definition with a working one?

In https://github.com/kubernetes/kubernetes/issues/109597 I mentioned another scenario to break an update of a statefulSet: requesting a non-existant secret.

Reply:
Put the known issue in the official doc without any workaround is not helping at all. Also I would like to know exactly what's the purpose of not updating the unhealthy pods? what's the point of keeping those "dead" pods anyway?